src/contexts/AuthContext.tsx

```diff
@@ -1,31 +1,164 @@
 
-export interface State {
+enum ActionTypes {
+  AUTH_INIT = 'auth/init',
+  AUTH_ERROR = 'auth/error',
+  AUTH_SUCCESS = 'auth/success',
+  AUTH_LOGOUT = 'auth/logout',
+}
+interface IAction {
+  type: ActionTypes;
+  payload?: AuthPayload;
+}
+
+interface IAuthProviderProps {
+  children: React.ReactNode;
+  value?: AuthState;
+}
+
+export interface AuthState {
   loading: boolean;
-  error: any;
+  error?: any;
   data: AuthPayload;
   isAuthenticated: boolean;
 }
 
-export interface AuthContextData extends State {
-  login: (email: string, password: string) => Promise<any>;
-  logout: () => void;
-}
-
-export const defaultState: State = {
+const defaultState: AuthState = {
   loading: false,
   error: null,
   data: {} as AuthPayload,
   isAuthenticated: false,
 };
 
-const defaultContext: AuthContextData = {
-  ...defaultState,
-  login: () => new Promise(() => null),
-  logout: () => null,
+const AuthContext = React.createContext<AuthState | undefined>(undefined);
+const AuthDispatchContext = React.createContext<React.Dispatch<IAction> | undefined>(
+  undefined
+);
+
+const authReducer = (state: AuthState, action: IAction): AuthState => {
+  switch (action.type) {
+    case ActionTypes.AUTH_INIT:
+      return {
+        ...state,
+        loading: true,
+        error: null,
+      };
+    case ActionTypes.AUTH_SUCCESS:
+      return {
+        ...state,
+        loading: false,
+        data: { ...action.payload } as AuthPayload,
+        isAuthenticated: true,
+      };
+    case ActionTypes.AUTH_ERROR:
+      return {
+        ...state,
+        loading: false,
+        error: { ...action.payload },
+      };
+    case ActionTypes.AUTH_LOGOUT:
+      return {
+        ...state,
+        data: {} as AuthPayload,
+        isAuthenticated: false,
+      };
+    default:
+      throw new Error('Not valid action');
+  }
 };
 
-const AuthContext = React.createContext<AuthContextData>(defaultContext);
+const AuthProvider = ({ children, ...props }: IAuthProviderProps) => {
+  const [state, dispatch] = useReducer(authReducer, defaultState);
+  const { data } = useQuery<authCache>(GET_LOCAL_AUTH);
+
+  useEffect(() => {
+    const user = auth.getUser();
+    if (user) {
+      dispatch({
+        type: ActionTypes.AUTH_SUCCESS,
+        payload: user,
+      });
+    }
+  }, []);
+
+  useEffect(() => {
+    if (data && data.auth && data.auth.tokenHasExpired) {
+      apolloClient.writeQuery<authCache>({
+        query: GET_LOCAL_AUTH,
+        data: {
+          auth: { __typename: 'auth', tokenHasExpired: false },
+        },
+      });
+      logout(dispatch);
+    }
+  }, [data]);
+
+  return (
+    <AuthContext.Provider value={state} {...props}>
+      <AuthDispatchContext.Provider value={dispatch}>
+        {children}
+      </AuthDispatchContext.Provider>
+    </AuthContext.Provider>
+  );
+};
+
+const useAuthState = () => {
+  const context = React.useContext(AuthContext);
+  if (context === undefined) {
```

Comment:
Is it possible that `context` could be `null` ?  I think it could be simplified `if(!context)`... same in line 122

Reply:
Not really, this is to protect the case when you try to access the context inside a component and the provider is not defined at the top level so it will always undefined

Reply:
Fair enough